### PR TITLE
Add Timeout when fetching state.json

### LINKED ISSUE #243: Add timeouts to HTTP /state.json requests
Currently these requests could hang for indeterminate time. https://github.com/mesosphere/mesos-dns/blob/master/records/generator.go#L106

### PR DESCRIPTION
**Pull request by sargun**
Fixes #243

### CODE REVIEW

File: config.json.sample [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {
   "zk": "zk://127.0.0.1:2181/mesos",
   "masters": ["127.0.0.1:5050"],
+  "StateTimeout": 300,
```
Comment (jdef) [inline config.json.sample]: probably better named `stateTimeoutSeconds` (lowercase, and express the unit)

File: records/config.go [hunk 2]
```diff
@@ -23,6 +23,8 @@ type Config struct {
 	// Timeout is the default connect/read/write timeout for outbound
 	// queries
 	Timeout int
+	// Timeout waiting for the master to return data from StateJson
```
Comment (jdef) [inline records/config.go]: indicate the unit here, please (seconds)

File: records/generator.go [hunk 3]
```diff
@@ -14,6 +14,8 @@ import (
 	"strconv"
 	"strings"
 
+	"time"
```
Comment (jdef) [inline records/generator.go]: nit: get rid of the newline above "time"

File: records/generator.go [hunk 4]
```diff
@@ -27,9 +29,17 @@ type rrs map[string][]string
...
-	SRVs     rrs
-	SlaveIPs map[string]string
+	As         rrs
+	SRVs       rrs
+	SlaveIPs   map[string]string
+	httpClient http.Client
+}
+
+// NewRecordGenerator returns a RecordGenerator that's been configured with a timeout.
+func NewRecordGenerator(httpTimeout time.Duration) *RecordGenerator {
```
Comment (jdef) [inline records/generator.go]: why pass in the timeout here? probably better off passing in a http.Client since there's likely to be other things that will want configuration in the future (e.g. SSL settings)

File: records/generator.go [hunk 5]
```diff
@@ -57,8 +67,8 @@ func (rg *RecordGenerator) ParseState(c Config, masters ...string) error {
 
 // Tries each master and looks for the leader
 // if no leader responds it errors
-func (rg *RecordGenerator) findMaster(masters ...string) (state.State, error) {
-	var sj state.State
+func (rg *RecordGenerator) findMaster(masters ...string) (*state.State, error) {
```
Comment (jdef) [inline records/generator.go]: why the conversion of `state.State` to `*state.State`?
- seems outside the scope of the goal of this PR
- pass-by-value isn't terribly expensive here since the bulk of the data is in slices, and those are copied by reference (cheap)

File: records/generator.go [hunk 6]
```diff
@@ -120,54 +128,56 @@ func (rg *RecordGenerator) loadFromMaster(ip string, port string) (sj state.Stat
...
 
-	client := &http.Client{}
-	resp, err := client.Do(req)
+	resp, err := rg.httpClient.Do(req)
 	if err != nil {
 		logging.Error.Println(err)
+		return nil, err
 	}
 
 	body, err := ioutil.ReadAll(resp.Body)
```
Comment (jdef) [inline records/generator.go]: should really have a `defer resp.Body.close()` here, before the `ReadAll` (and get rid of the one on line 148)

File: records/generator.go [hunk 7]
```diff
@@ -108,9 +115,10 @@ func (rg *RecordGenerator) findMaster(masters ...string) (state.State, error) {
 }
 
 // Loads state.json from mesos master
-func (rg *RecordGenerator) loadFromMaster(ip string, port string) (sj state.State) {
+func (rg *RecordGenerator) loadFromMaster(ip string, port string) (*state.State, error) {
```
Comment (jdef) [inline records/generator.go]: if you're going to return error's here, why bother logging them in this func? leave that up to the caller (and if the caller doesn't log (but also just passes the error back) then it's the grand-caller's job to log, so on and so forth).

File: records/generator.go [hunk 8]
```diff
@@ -198,7 +208,7 @@ func hostToIP4(hostname string) (string, bool) {
 }
 
 // InsertState transforms a StateJSON into RecordGenerator RRs
-func (rg *RecordGenerator) InsertState(sj state.State, domain, ns, listener string, masters, ipSources []string, spec labels.Func) error {
+func (rg *RecordGenerator) InsertState(sj *state.State, domain, ns, listener string, masters, ipSources []string, spec labels.Func) error {
```
Comment (jdef) [inline records/generator.go]: ditto .. no need for the `*state.State` change everywhere

File: records/generator_test.go [hunk 9]
```diff
@@ -7,6 +7,12 @@ import (
 	"testing"
 	"testing/quick"
 
+	"net"
```
Comment (jdef) [inline records/generator_test.go]: nit: these imports should be merged with those above

File: resolver/resolver.go [hunk 10]
```diff
@@ -151,7 +154,7 @@ func (res *Resolver) Reload() {
 		atomic.StoreUint32(&res.config.SOASerial, timestamp)
 		res.rs = &t
 	} else {
-		logging.VeryVerbose.Println("Warning: master not found; keeping old DNS state")
+		logging.VeryVerbose.Printf("Warning: Error generating records: %v; keeping old DNS state", err)
```
Comment (jdef) [inline resolver/resolver.go]: why do we log an error condition as VeryVerbose?

File: records/generator.go (hunk 6 above)
Comment (jdef) [inline records/generator.go]: e.g.
```go
defer resp.Body.Close()
body, err := ioutil.ReadAll(resp.Body)
```

**Comment by jdef** [comment]
initial review complete

File: records/generator.go (hunk 5 above)
Comment (sargun) [inline records/generator.go]: It's primarily to be able to provide a nil (Optional). I'd rather mesos-dns fail in an obvious way if somehow error checking fails, as opposed to returning an empty state, and the entire zone emptying out.

File: records/generator.go (hunk 8 above)
Comment (sargun) [inline records/generator.go]: See above.

File: records/generator.go (hunk 4 above)
Comment (sargun) [inline records/generator.go]: I'm still investigating cases where we might want to blow away the httpclient entirely (stale connections, etc..). Originally, I had it where if HTTP timeout was encountered, it'd blow away the old client. I'm not sure how legitimate this concern is, hence I didn't proceed.

File: config.json.sample (hunk 1 above)
Comment (sargun) [inline config.json.sample]: Okay. I'm going to set the field as StateTimeoutSeconds in JSON, rather than using a JSON Tag.

File: records/generator.go (hunk 7 above)
Comment (sargun) [inline records/generator.go]: Got rid of the error logging.

File: records/config.go [hunk 11]
```diff
@@ -67,26 +69,27 @@ type Config struct {
 // NewConfig return the default config of the resolver
 func NewConfig() Config {
 	return Config{
```
Comment (sargun) [inline records/config.go]: Gofmt shifted everything over due to a longer field

File: records/generator.go (hunk 5 above)
Comment (jdef) [inline records/generator.go]: returning error objects upon an error condition is idiomatic Go. if err == nil then there's no error. having two possible error states to check (error object, or state == nil) is confusing and not intuitive.

Comment (sargun) [inline records/generator.go]: I think my reasoning here was state == nil iff err != nil. If someone is checking the err condition, and bailing out, there should be no need for them to ever check whether state == nil. I see the benefit as if the value is somehow dereferenced, because error checking went awry, it'll crash. From my understanding, mesos-dns is typically run under a supervisor, which will restart it. I see this as a better alternative than loading in an incorrect state. I see this as a variant of crash-only computing.

If you look at the behaviour in the Go standard library, it appears to be the same: https://play.golang.org/p/aUUoHoNXnM. 

Although, I've noticed in our project, we've followed the semantic of preferring to load the zero value. If we'd like to continue that, I don't think it's detrimental.

File: config.json.sample [hunk 12]
```diff
@@ -1,6 +1,7 @@
 {
   "zk": "zk://127.0.0.1:2181/mesos",
   "masters": ["127.0.0.1:5050"],
+  "stateTimeoutSeconds": 300,
```
Comment (sargun) [inline config.json.sample]: One of the questions I have is why we didn't use strings, and time.Duration / time.ParseDuration? @jdef do you know?

File: records/generator.go (hunk 5 above)
Comment (jdef) [inline records/generator.go]: It really depends on the API. For example, the io.Reader api tells you to
check for data, regardless of whether an error is returned.

We don't need complicated API semantics here because we don't have
different types of errors to handle in different ways. If err != nil then I
think we shouldn't trust the returned state -- end of story.

I don't hate the change to *state.State -- I just don't think it's
necessary.

File: config.json.sample (hunk 12 above)
Comment (jdef) [inline config.json.sample]: I can't think of a good reason, probably because there isn't one.

File: records/generator.go (hunk 5 above)
Comment (sargun) [inline records/generator.go]: Understood.

File: config.json.sample (hunk 12 above)
Comment (sargun) [inline config.json.sample]: I created a related bug: mesos-dns #342

File: records/generator.go [hunk 13]
```diff
@@ -152,22 +163,21 @@ func (rg *RecordGenerator) loadWrap(ip string, port string) (state.State, error)
...
-
+	sj, err = rg.loadFromMaster(ip, port)
+	if err != nil {
+		return state.State{}, err
+	}
 	if rip := leaderIP(sj.Leader); rip != ip {
 		logging.VeryVerbose.Println("Warning: master changed to " + ip)
-		sj = rg.loadFromMaster(rip, port)
+		sj, err = rg.loadFromMaster(rip, port)
+		if err != nil {
```
Comment (jdef) [inline records/generator.go]: why do you need this `if`? can't you just `return sj, err`?

File: resolver/resolver.go [hunk 14]
```diff
@@ -140,7 +142,8 @@ func (res *Resolver) SetMasters(masters []string) {
 // Reload triggers a new state load from the configured mesos masters.
 // This method is not goroutine-safe.
 func (res *Resolver) Reload() {
-	t := records.RecordGenerator{}
+	var t records.RecordGenerator
```
Comment (jdef) [inline resolver/resolver.go]: this is weird. just say `t := records.NewRecordGenerator(...)`

File: records/generator.go [hunk 15]
```diff
@@ -120,28 +128,31 @@ func (rg *RecordGenerator) loadFromMaster(ip string, port string) (sj state.Stat
...
 
-	client := &http.Client{}
-	resp, err := client.Do(req)
+	resp, err := rg.httpClient.Do(req)
 	if err != nil {
 		logging.Error.Println(err)
+		return state.State{}, err
 	}
 
+	defer util.IgnoreError(resp.Body.Close)
```
Comment (jdef) [inline records/generator.go]: this is an interesting idea. we normally don't care about close errors. not sure why we need to be explicit here with IgnoreError()

File: util/util.go [hunk 16]
```diff
@@ -52,3 +52,22 @@ func logPanic(r interface{}) {
...
+// IgnoreError Calls an ErrorFunction, and ignores the result.
+// This allows us to be more explicit when there is no error
+// handling to be done, for example in defers
+func IgnoreError(f ErrorFunction) {
+	_ = f()
+}
+
+// PanicOnError Calls an ErrorFunction, and panics if the result
+// isn't nil.
+func PanicOnError(f ErrorFunction) {
```
Comment (jdef) [inline util/util.go]: nothing uses this, except for a test case. are there bigger plans for incorporating panics?

File: records/generator.go (hunk 4 above)
Comment (jdef) [inline records/generator.go]: I think it's recommended to reuse http.Client instances if possible. People that haven't done so have ended up with connection leaks

Comment (sargun) [inline records/generator.go]: We weren't reusing the HTTP client before. Is there any reason that was the case? I personally don't know, but, if you're confident it's safe to do it, I'll make the change to have it passed in.

Comment (jdef) [inline records/generator.go]: hm .. let's re-use it in a separate PR

File: records/generator.go (hunk 15 above)
Comment (sargun) [inline records/generator.go]: This is mostly a matter of dealing with the linters. Related: https://github.com/mesosphere/mesos-dns/issues/223 -- I don't think it's a good idea to ignore those errors completely. I think that being explicit about ignoring errors is valuable, and as laid out in the issue, we'd want to handle R/W objects differently.

File: records/generator.go (hunk 4 above)
Comment (sargun) [inline records/generator.go]: Will do - it's a simple change, but I think it should be looked at.

File: records/generator_test.go [hunk 17]
```diff
@@ -164,6 +169,7 @@ func testRecordGenerator(t *testing.T, spec labels.Func, ipSources []string) Rec
 	if err != nil {
 		t.Fatal(err)
 	} else if err = json.Unmarshal(b, &sj); err != nil {
+		t.Log("err: ", err)
```
Comment (jdef) [inline records/generator_test.go]: this is strange: we're already logging the error w/ t.Fatal(), why the additional t.Log()?

File: records/generator_test.go [hunk 18]
```diff
@@ -267,3 +273,26 @@ func TestHashString(t *testing.T) {
...
+		notify := w.(http.CloseNotifier).CloseNotify()
+		<-notify
+	}
+	server := httptest.NewServer(http.HandlerFunc(sleepForeverHandler))
+	defer server.Close()
+
+	rg := NewRecordGenerator(500 * time.Millisecond)
+	host, port, err := net.SplitHostPort(server.Listener.Addr().String())
+	_, err = rg.loadFromMaster(host, port)
+	netErr := (err).(*url.Error).Err.(net.Error)
```
Comment (jdef) [inline records/generator_test.go]: if might make for a more readable failure if there some additional assertions here:
- `if err == nil { f.Fatalf("expected error because ...") }`
- `urlErr, ok := ...; if !ok { t.Fatalf("expected url.Error instead of %#+v", err) }`

... etc

**Comment by jdef** [comment]
nits: otherwise lgtm

File: records/generator_test.go (hunk 17 above)
Comment (sargun) [inline records/generator_test.go]: I have no idea how this snuck in. Getting rid of it.

**Comment by jdef** [comment]
:shipit:
